core/services/keeper/upkeep_executer.go

```diff
@@ -24,40 +24,50 @@ const (
 )
 
 func NewUpkeepExecutor(
+	job job.Job,
 	db *gorm.DB,
 	ethClient eth.Client,
+	headRelayer *services.HeadRelayer,
 ) *UpkeepExecutor {
 	return &UpkeepExecutor{
-		blockHeight:    atomic.NewInt64(0),
-		wgDone:         sync.WaitGroup{},
+		chStop:         make(chan struct{}),
 		ethClient:      ethClient,
-		orm:            NewORM(db),
 		executionQueue: make(chan struct{}, executionQueueSize),
-		chStop:         make(chan struct{}),
-		chSignalRun:    make(chan struct{}, 1),
+		headRelayer:    headRelayer,
+		job:            job,
+		mailbox:        utils.NewMailbox(1),
+		orm:            NewORM(db),
+		wgDone:         sync.WaitGroup{},
 		StartStopOnce:  utils.StartStopOnce{},
 	}
 }
 
-// UpkeepExecutor fulfills HeadTrackable interface
-var _ store.HeadTrackable = (*UpkeepExecutor)(nil)
+// UpkeepExecutor fulfills Service and HeadRelayable interfaces
+var _ job.Service = (*UpkeepExecutor)(nil)
+var _ services.HeadRelayable = (*UpkeepExecutor)(nil)
 
 type UpkeepExecutor struct {
-	blockHeight *atomic.Int64
-	wgDone      sync.WaitGroup
-	ethClient   eth.Client
-	orm         ORM
-
-	executionQueue chan struct{}
 	chStop         chan struct{}
-	chSignalRun    chan struct{}
-
+	ethClient      eth.Client
+	executionQueue chan struct{}
+	headRelayer    *services.HeadRelayer
+	job            job.Job
+	mailbox        *utils.Mailbox
+	orm            ORM
+	wgDone         sync.WaitGroup
 	utils.StartStopOnce
 }
 
 func (executor *UpkeepExecutor) Start() error {
 	return executor.StartOnce("UpkeepExecutor", func() error {
 		go executor.run()
+		unsubscribe := executor.headRelayer.Subscribe(executor)
+		executor.wgDone.Add(1)
```

Comment:
Probably just calling it `executor.wg` is clear enough

Reply:
I like the consistency of calling it the same name across different services that implement this pattern. And using `wgDone` differentiates it from other local `wg` variables in other places in the code base.